csharp/src/SeedLang/Visualization/EventData.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021-2022 The SeedV Lab.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace SeedLang.Visualization {
+  public enum VariableType {
+    Global,
+    Local,
+    Upvalue,
+  }
+
+  public enum LValueType {
+    ElementOfContainer,
+    Variable,
+  }
+
+  public enum RValueType {
+    ElementOfContainer,
+    TemporaryValue,
+    Variable,
+  }
+
+  // The LValue class. It could be a variable or an element of containers. The Keys field is null if
+  // it's a variable.
```

Comment:
For a better interface for this class, it's also good to provide properties to let the client code easily determine the type. E.g.,

```
public bool IsVariable => key is null;
public bool IsElement => !IsVariable;
```

Reply:
Done.